File: swift/Tests/Bandwidth/BandwidthFixtureTests.swift
// SPDX-License-Identifier: MIT

import Foundation
import XCTest
@testable import AetherNetProtocol

/// Drives the Swift ABMF SDK through the cross-language conformance corpus at
/// `tests/cross-language/bandwidth-fixtures.json`. Every AetherNet SDK drives the
/// SAME corpus and MUST produce identical results — this is the oracle that proves
/// numeric parity across all 8 languages. Mirrors the C# reference driver in
/// `AetherNet.Core.Tests/Bandwidth/BandwidthFixtureTests.cs` case-for-case.
///
/// Integer/string/enum fields are asserted EXACTLY; floating-point fields
/// (srttMs, rttVarMs, rtPropMs, lossRate) within `toleranceAbs` of the expected.
///
/// The estimator/director are `actor`s, so every drive method is `async` and uses
/// `await` to reach `currentSample`, `applyPhyHint`, `recommendTransport`, etc.
final class BandwidthFixtureTests: XCTestCase {

    // MARK: - Corpus model

    private struct ProbeAckCase: Decodable {
        let name: String
        let senderSendUs: Int64
        let receiverReceiveUs: Int64
        let receiverSendUs: Int64
        let senderReceiveUs: Int64
        let probeBytes: Int32
        let expectRttUs: Int64
        let expectForwardOwdUs: Int64
    }

    private struct RtoCase: Decodable {
        let name: String
        let srttMs: Double
        let rttVarMs: Double
        let expectRtoMs: Double
    }

    private struct PhyCapCase: Decodable {
        let name: String
        let rssiDbm: Int
        let expectCapBps: Int64
    }

    /// One operation in an estimator drive. Field set depends on `op`, so every
    /// payload field is optional (mirrors the C# `TryGetProperty` switch).
    private struct EstimatorOp: Decodable {
        let op: String
        let bytes: Int?
        let sendUs: Int64?
        let deliverUs: Int64?
        let rssiDbm: Int?
        let btlBwBps: Int64?
        let rtPropMs: Double?
        let confidence: String?
    }

    /// Expected sample fields. All optional — a case asserts only the fields it
    /// declares (mirrors the C# `TryGetProperty` blocks).
    private struct EstimatorExpect: Decodable {
        let btlBwBps: Int64?
        let effectiveBps: Int64?
        let availableBps: Int64?
        let bdpBytes: Int64?
        let phyCapBps: Int64?
        let confidence: String?
        let srttMs: Double?
        let rttVarMs: Double?
        let rtPropMs: Double?
        let lossRate: Double?
    }

    private struct EstimatorCase: Decodable {
        let name: String
        let transport: String
        let maxBps: Int64
        let ops: [EstimatorOp]
        let expect: EstimatorExpect
    }

    private struct DirectorGossip: Decodable {
        let peerUhid: String
        let transport: String
        let btlBwBps: Int64
        let rtPropUs: Int64
        let confidence: String
    }

    private struct DirectorRecommend: Decodable {
        let peerUhid: String
        let payloadBytes: Int64
    }

    private struct DirectorCase: Decodable {
        let name: String
        let register: [String]
        let gossips: [DirectorGossip]
        let recommend: DirectorRecommend
        /// JSON null → nil (no recommendation expected).
        let expectTransport: String?
    }

    private struct Corpus: Decodable {
        let toleranceAbs: Double
        let probeAck: [ProbeAckCase]
        let rto: [RtoCase]
        let phyCap: [PhyCapCase]
        let estimator: [EstimatorCase]
        let director: [DirectorCase]
    }

    // MARK: - Fixture loader

    /// Locate `tests/cross-language/bandwidth-fixtures.json` by walking up from this
    /// source file's directory (`#file`) to the repo root — independent of CWD, the
    /// same parent-traversal idiom as the URI CrossLanguageFixtureTests.
    private func loadCorpus() throws -> Corpus {
        var url = URL(fileURLWithPath: #file).deletingLastPathComponent()
        for _ in 0..<10 {
            let candidate = url
                .appendingPathComponent("tests")
                .appendingPathComponent("cross-language")
                .appendingPathComponent("bandwidth-fixtures.json")
            if FileManager.default.fileExists(atPath: candidate.path) {
                let data = try Data(contentsOf: candidate)
                return try JSONDecoder().decode(Corpus.self, from: data)
            }
            let parent = url.deletingLastPathComponent()
            if parent.path == url.path { break }
            url = parent
        }
        XCTFail("Could not locate tests/cross-language/bandwidth-fixtures.json walking up from \(#file)")
        throw CocoaError(.fileNoSuchFile)
    }

    private func parseConfidence(_ s: String) -> BandwidthConfidence {
        switch s {
        case "None":   return .none
        case "Low":    return .low
        case "Medium": return .medium
        case "High":   return .high
        default:
            XCTFail("bad confidence \(s)")
            return .none
        }
    }

    /// Round seconds → microseconds the same way the corpus was generated.
    private func toMicros(_ seconds: TimeInterval) -> Int64 {
        Int64((seconds * 1_000_000.0).rounded())
    }

    // MARK: - probeAck

    func testProbeAck_RttAndOwd_Exact() throws {
        let corpus = try loadCorpus()
        for f in corpus.probeAck {
            let ack = BandwidthProbeAck(
                sequence: 1,
                senderSendUs: f.senderSendUs,
                receiverReceiveUs: f.receiverReceiveUs,
                receiverSendUs: f.receiverSendUs,
                senderReceiveUs: f.senderReceiveUs,
                probeBytes: f.probeBytes
            )

            XCTAssertEqual(toMicros(ack.rtt), f.expectRttUs,
                "[\(f.name)] rtt µs mismatch")
            XCTAssertEqual(toMicros(ack.forwardOwd), f.expectForwardOwdUs,
                "[\(f.name)] forwardOwd µs mismatch")
        }
    }

    // MARK: - rto

    func testRto_Clamped_MatchesRfc6298() throws {
        let corpus = try loadCorpus()
        for f in corpus.rto {
            let sample = BandwidthSample(
                transportName: "T",
                btlBwBps:      1_000_000,
                availableBps:  900_000,
                bdpBytes:      1000,
                srtt:          f.srttMs / 1000.0,
                rttVar:        f.rttVarMs / 1000.0,
                rtProp:        0.010,
                lossRate:      0.0,
                phyCapBps:     0,
                confidence:    .high,
                measuredAt:    Date()
            )

            XCTAssertEqual(sample.rto * 1000.0, f.expectRtoMs, accuracy: 0.1,
                "[\(f.name)] rto ms mismatch")
        }
    }

    // MARK: - phyCap

    func testPhyCap_FromRssi_Exact() async throws {
        let corpus = try loadCorpus()
        for f in corpus.phyCap {
            let est = BandwidthEstimator(transportName: "T", maxBandwidthBps: 10_000_000_000)
            await est.applyPhyHint(rssiDbm: f.rssiDbm)
            let sample = await est.currentSample
            XCTAssertEqual(sample.phyCapBps, f.expectCapBps,
                "[\(f.name)] phyCapBps mismatch")
        }
    }

    // MARK: - estimator

    func testEstimator_DrivesToExpectedSample() async throws {
        let corpus = try loadCorpus()
        let tol = corpus.toleranceAbs

        for f in corpus.estimator {
            let est = BandwidthEstimator(transportName: f.transport, maxBandwidthBps: f.maxBps)

            for op in f.ops {
                switch op.op {
                case "delivery":
                    await est.recordDelivery(
                        bytes: op.bytes!,
                        sendUs: op.sendUs!,
                        deliverUs: op.deliverUs!
                    )
                case "loss":
                    await est.recordLoss(bytes: op.bytes!)
                case "phyHint":
                    await est.applyPhyHint(rssiDbm: op.rssiDbm!)
                case "gossip":
                    // JSON rtPropMs is milliseconds; warmFromGossip takes seconds.
                    await est.warmFromGossip(
                        btlBwBps: op.btlBwBps!,
                        rtProp: op.rtPropMs! / 1000.0,
                        confidence: parseConfidence(op.confidence!)
                    )
                default:
                    XCTFail("[\(f.name)] unknown op \(op.op)")
                }
            }

            let s = await est.currentSample
            let exp = f.expect

            // Integer / enum fields — exact.
            if let v = exp.btlBwBps     { XCTAssertEqual(s.btlBwBps, v,     "[\(f.name)] btlBwBps") }
            if let v = exp.effectiveBps { XCTAssertEqual(s.effectiveBps, v, "[\(f.name)] effectiveBps") }
            if let v = exp.availableBps { XCTAssertEqual(s.availableBps, v, "[\(f.name)] availableBps") }
            if let v = exp.bdpBytes     { XCTAssertEqual(s.bdpBytes, v,     "[\(f.name)] bdpBytes") }
            if let v = exp.phyCapBps    { XCTAssertEqual(s.phyCapBps, v,    "[\(f.name)] phyCapBps") }
            if let v = exp.confidence   { XCTAssertEqual(s.confidence, parseConfidence(v), "[\(f.name)] confidence") }

            // Float fields — tolerance (TimeInterval is seconds → compare in ms).
            if let v = exp.srttMs   { XCTAssertEqual(s.srtt * 1000.0, v,   accuracy: tol, "[\(f.name)] srttMs") }
            if let v = exp.rttVarMs { XCTAssertEqual(s.rttVar * 1000.0, v, accuracy: tol, "[\(f.name)] rttVarMs") }
            if let v = exp.rtPropMs { XCTAssertEqual(s.rtProp * 1000.0, v, accuracy: tol, "[\(f.name)] rtPropMs") }
            if let v = exp.lossRate { XCTAssertEqual(s.lossRate, v,        accuracy: tol, "[\(f.name)] lossRate") }
        }
    }

    // MARK: - director

    func testDirector_RecommendsExpectedTransport() async throws {
        let corpus = try loadCorpus()

        for f in corpus.director {
            let director = BandwidthDirector()

            // Register one estimator per declared transport. Generous maxBps so the
            // PHY default does not cap gossip-seeded values.
            for t in f.register {
                await director.register(BandwidthEstimator(transportName: t, maxBandwidthBps: 10_000_000_000))
            }

            for g in f.gossips {
                await director.applyGossip(BandwidthGossipPayload(
                    peerUhid:      g.peerUhid,
                    transportName: g.transport,
                    btlBwBps:      g.btlBwBps,
                    rtPropUs:      g.rtPropUs,
                    confidence:    parseConfidence(g.confidence),
                    measuredAt:    Date()
                ))
            }

            let result = await director.recommendTransport(
                peerUhid: f.recommend.peerUhid,
                payloadBytes: f.recommend.payloadBytes
            )

            if let expected = f.expectTransport {
                XCTAssertEqual(result, expected, "[\(f.name)] recommended transport mismatch")
            } else {
                XCTAssertNil(result, "[\(f.name)] expected no recommendation")
            }
        }
    }
}

File: tests/AetherNet.Core.Tests/Bandwidth/BandwidthFixtureTests.cs
// SPDX-License-Identifier: MIT

using System.IO;
using System.Linq;
using System.Text.Json;
using AetherNet.Bandwidth;
using AetherNet.Transport.Bandwidth;
using Xunit;

namespace AetherNet.Core.Tests.Bandwidth;

/// <summary>
/// Drives the C# reference through the cross-language ABMF corpus at
/// <c>tests/cross-language/bandwidth-fixtures.json</c>. Every other AetherNet SDK
/// drives the SAME corpus and MUST produce identical results. This is the oracle
/// that proves numeric parity across all 8 languages — without it, "identical by
/// construction" is unverified (and the bdpBonus divergence proved construction
/// is not enough).
/// </summary>
public class BandwidthFixtureTests
{
    private static readonly JsonElement Root = LoadCorpus();
    private static readonly double Tol = Root.GetProperty("toleranceAbs").GetDouble();

    private static JsonElement LoadCorpus()
    {
        var dir = new DirectoryInfo(AppContext.BaseDirectory);
        while (dir is not null)
        {
            var candidate = Path.Combine(dir.FullName, "tests", "cross-language", "bandwidth-fixtures.json");
            if (File.Exists(candidate))
                return JsonDocument.Parse(File.ReadAllText(candidate)).RootElement.Clone();
            dir = dir.Parent;
        }
        throw new FileNotFoundException("bandwidth-fixtures.json not found walking up from " + AppContext.BaseDirectory);
    }

    private static BandwidthConfidence ParseConfidence(string s) => s switch
    {
        "None" => BandwidthConfidence.None,
        "Low" => BandwidthConfidence.Low,
        "Medium" => BandwidthConfidence.Medium,
        "High" => BandwidthConfidence.High,
        _ => throw new ArgumentException($"bad confidence {s}"),
    };

    // ── probeAck ──────────────────────────────────────────────────────────────

    public static IEnumerable<object[]> ProbeAckCases() =>
        Root.GetProperty("probeAck").EnumerateArray().Select(e => new object[] { e.GetProperty("name").GetString()!, e });

    [Theory]
    [MemberData(nameof(ProbeAckCases))]
    public void ProbeAck_RttAndOwd_Exact(string name, JsonElement f)
    {
        _ = name;
        var ack = new BandwidthProbeAck(
            (uint)1,
            f.GetProperty("senderSendUs").GetInt64(),
            f.GetProperty("receiverReceiveUs").GetInt64(),
            f.GetProperty("receiverSendUs").GetInt64(),
            f.GetProperty("senderReceiveUs").GetInt64(),
            f.GetProperty("probeBytes").GetInt32());

        Assert.Equal(f.GetProperty("expectRttUs").GetInt64(), (long)ack.Rtt.TotalMicroseconds);
        Assert.Equal(f.GetProperty("expectForwardOwdUs").GetInt64(), (long)ack.ForwardOwd.TotalMicroseconds);
    }

    // ── rto ───────────────────────────────────────────────────────────────────

    public static IEnumerable<object[]> RtoCases() =>
        Root.GetProperty("rto").EnumerateArray().Select(e => new object[] { e.GetProperty("name").GetString()!, e });

    [Theory]
    [MemberData(nameof(RtoCases))]
    public void Rto_Clamped_MatchesRfc6298(string name, JsonElement f)
    {
        _ = name;
        var sample = new BandwidthSample(
            "T", 1_000_000, 900_000, 1000,
            TimeSpan.FromMilliseconds(f.GetProperty("srttMs").GetDouble()),
            TimeSpan.FromMilliseconds(f.GetProperty("rttVarMs").GetDouble()),
            TimeSpan.FromMilliseconds(10), 0.0, 0L, BandwidthConfidence.High, DateTimeOffset.UtcNow);

        Assert.Equal(f.GetProperty("expectRtoMs").GetDouble(), sample.Rto.TotalMilliseconds, precision: 1);
    }

    // ── phyCap ─────────────────────────────────────────────────────────────────

    public static IEnumerable<object[]> PhyCapCases() =>
        Root.GetProperty("phyCap").EnumerateArray().Select(e => new object[] { e.GetProperty("name").GetString()!, e });

    [Theory]
    [MemberData(nameof(PhyCapCases))]
    public void PhyCap_FromRssi_Exact(string name, JsonElement f)
    {
        _ = name;
        var e = new BandwidthEstimator("T", 10_000_000_000L);
        e.ApplyPhyHint(f.GetProperty("rssiDbm").GetInt32());
        Assert.Equal(f.GetProperty("expectCapBps").GetInt64(), e.CurrentSample.PhyCapBps);
    }

    // ── estimator ──────────────────────────────────────────────────────────────

    public static IEnumerable<object[]> EstimatorCases() =>
        Root.GetProperty("estimator").EnumerateArray().Select(e => new object[] { e.GetProperty("name").GetString()!, e });

    [Theory]
    [MemberData(nameof(EstimatorCases))]
    public void Estimator_DrivesToExpectedSample(string name, JsonElement f)
    {
        _ = name;
        var e = new BandwidthEstimator(f.GetProperty("transport").GetString()!, f.GetProperty("maxBps").GetInt64());

        foreach (var op in f.GetProperty("ops").EnumerateArray())
        {
            switch (op.GetProperty("op").GetString())
            {
                case "delivery":
                    e.RecordDelivery(op.GetProperty("bytes").GetInt32(),
                                     op.GetProperty("sendUs").GetInt64(),
                                     op.GetProperty("deliverUs").GetInt64());
                    break;
                case "loss":
                    e.RecordLoss(op.GetProperty("bytes").GetInt32());
                    break;
                case "phyHint":
                    e.ApplyPhyHint(op.GetProperty("rssiDbm").GetInt32());
                    break;
                case "gossip":
                    e.WarmFromGossip(op.GetProperty("btlBwBps").GetInt64(),
                                     TimeSpan.FromMilliseconds(op.GetProperty("rtPropMs").GetDouble()),
                                     ParseConfidence(op.GetProperty("confidence").GetString()!));
                    break;
                default: throw new ArgumentException("unknown op");
            }
        }

        var s = e.CurrentSample;
        var exp = f.GetProperty("expect");

        // Integer / enum fields — exact.
        if (exp.TryGetProperty("btlBwBps", out var v1))    Assert.Equal(v1.GetInt64(), s.BtlBwBps);
        if (exp.TryGetProperty("effectiveBps", out var v2)) Assert.Equal(v2.GetInt64(), s.EffectiveBps);
        if (exp.TryGetProperty("availableBps", out var v3)) Assert.Equal(v3.GetInt64(), s.AvailableBps);
        if (exp.TryGetProperty("bdpBytes", out var v4))    Assert.Equal(v4.GetInt64(), s.BdpBytes);
        if (exp.TryGetProperty("phyCapBps", out var v5))   Assert.Equal(v5.GetInt64(), s.PhyCapBps);
        if (exp.TryGetProperty("confidence", out var v6))  Assert.Equal(ParseConfidence(v6.GetString()!), s.Confidence);

        // Float fields — tolerance.
        if (exp.TryGetProperty("srttMs", out var f1))   Assert.Equal(f1.GetDouble(), s.Srtt.TotalMilliseconds, Tol);
        if (exp.TryGetProperty("rttVarMs", out var f2)) Assert.Equal(f2.GetDouble(), s.RttVar.TotalMilliseconds, Tol);
        if (exp.TryGetProperty("rtPropMs", out var f3)) Assert.Equal(f3.GetDouble(), s.RtProp.TotalMilliseconds, Tol);
        if (exp.TryGetProperty("lossRate", out var f4)) Assert.Equal(f4.GetDouble(), s.LossRate, Tol);
    }

    // ── director ───────────────────────────────────────────────────────────────

    public static IEnumerable<object[]> DirectorCases() =>
        Root.GetProperty("director").EnumerateArray().Select(e => new object[] { e.GetProperty("name").GetString()!, e });

    [Theory]
    [MemberData(nameof(DirectorCases))]
    public void Director_RecommendsExpectedTransport(string name, JsonElement f)
    {
        _ = name;
        var director = new BandwidthDirector();

        // Register one estimator per declared transport. Use a generous maxBps so
        // the PHY default does not cap the gossip-seeded values.
        foreach (var t in f.GetProperty("register").EnumerateArray())
            director.Register(new BandwidthEstimator(t.GetString()!, 10_000_000_000L));

        foreach (var g in f.GetProperty("gossips").EnumerateArray())
        {
            director.ApplyGossip(new BandwidthGossipPayload(
                g.GetProperty("peerUhid").GetString()!,
                g.GetProperty("transport").GetString()!,
                g.GetProperty("btlBwBps").GetInt64(),
                g.GetProperty("rtPropUs").GetInt64(),
                ParseConfidence(g.GetProperty("confidence").GetString()!),
                DateTimeOffset.UtcNow));
        }

        var rec = f.GetProperty("recommend");
        var result = director.RecommendTransport(rec.GetProperty("peerUhid").GetString()!, rec.GetProperty("payloadBytes").GetInt64());

        var expectEl = f.GetProperty("expectTransport");
        if (expectEl.ValueKind == JsonValueKind.Null)
            Assert.Null(result);
        else
            Assert.Equal(expectEl.GetString(), result);
    }
}
